File: src/pooh3/mobi/app/a01_clock/model/TestClock.java
package pooh3.mobi.app.a01_clock.model;

import pooh3.mobi.app.a01_clock.application.GetClockHandsDegree;
import sun.jvm.hotspot.utilities.Assert;

import java.util.ArrayList;
import java.util.List;

public class TestClock {

    private List<Fixture>  longHandFixtures = new ArrayList<>();
    private List<Fixture>  shortHandFixtures = new ArrayList<>();
    private List<Fixture>  shortHandMinFixtures = new ArrayList<>();

    {
        longHandFixtures.add(new Fixture(0, 0));
        longHandFixtures.add(new Fixture(15, 90));
        longHandFixtures.add(new Fixture(30, 180));
        longHandFixtures.add(new Fixture(45, 270));
//        longHandFixtures.add(new Fixture(60, 360)); // exception

        shortHandFixtures.add(new Fixture(0, 0));
        shortHandFixtures.add(new Fixture(3, 90));
        shortHandFixtures.add(new Fixture(6, 180));
        shortHandFixtures.add(new Fixture(9, 270));
        shortHandFixtures.add(new Fixture(12, 0));
        shortHandFixtures.add(new Fixture(15, 90));
        shortHandFixtures.add(new Fixture(18, 180));
        shortHandFixtures.add(new Fixture(21, 270));

        shortHandMinFixtures.add(new Fixture(0, 0));
        shortHandMinFixtures.add(new Fixture(3, 90));
        shortHandMinFixtures.add(new Fixture(6, 180));
        shortHandMinFixtures.add(new Fixture(9, 90));
        shortHandMinFixtures.add(new Fixture(12, 0));
        shortHandMinFixtures.add(new Fixture(15, 90));
        shortHandMinFixtures.add(new Fixture(18, 180));
        shortHandMinFixtures.add(new Fixture(21, 90));
    }

    public void test() throws Exception {
        setUp();
        aLongHandTest_minute_to_degree();
        aShortHandTest_hour_to_degree();
        aShortHandTest_hour_to_min_degree();
        hour10minute10_answer115degree();
    }

    public void setUp() {
    }

    public void aLongHandTest_minute_to_degree() throws Exception {
        for (Fixture longHandFixture : longHandFixtures) {

            int degree =
                    LongHand.of(longHandFixture.aTime).degree().val;

            System.out.println(
                    "(hand-degree (long-hand " + longHandFixture.aTime
                            + ")) -> " + degree
                            + " == " + longHandFixture.aDegree);

            Assert.that(degree == longHandFixture.aDegree,
                    longHandFixture.aTime +" actual:"
                            + degree + " expected:" + longHandFixture.aDegree);
        }
    }

    public void aShortHandTest_hour_to_degree() throws Exception {
        for (Fixture shortHandFixture : shortHandFixtures) {

            int degree =
                    ShortHand.of(shortHandFixture.aTime)
                            .degree(LongHand.of(0)).val;

            System.out.println(
                    "(hand-degree (short-hand " + shortHandFixture.aTime
                            + ")) -> " + degree
                            + " == " + shortHandFixture.aDegree);

            Assert.that(degree == shortHandFixture.aDegree,
                    shortHandFixture.aTime +" actual:"
                            + degree + " expected:" + shortHandFixture.aDegree);
        }
    }

    public void aShortHandTest_hour_to_min_degree() throws Exception {
        for (Fixture shortHandFixture : shortHandMinFixtures) {

            Degree minDegree =
                    ShortHand.of(shortHandFixture.aTime).degree(LongHand.of(0))
                            .minusMin(Degree.of(0));

            System.out.println(
                    "(degree-min (hand-degree (short-hand " + shortHandFixture.aTime
                            + ")) -> " + minDegree.val
                            + " == " + shortHandFixture.aDegree);

            Assert.that(minDegree.val == shortHandFixture.aDegree,
                    shortHandFixture.aTime + " actual:"
                            + minDegree.val + " expected:" + shortHandFixture.aDegree);
        }
    }

    public void hour10minute10_answer115degree() {
        String c = "10:10";
        DegreeCalcService dcs = new DegreeCalcService();
        Degree degree = new GetClockHandsDegree(dcs).execute(c);

        System.out.println(
                "(degree-min (clock-degree (clock-str \"" + c
                        + "\")) -> " + degree.val
                        + " == " + 115);

        Assert.that(degree.val == 115,
                c + " actual:"
                        + degree.val + " expected:" + 115);
    }

    static class Fixture {
        public final int aTime;
        public final int aDegree;

        Fixture(int aTime, int aDegree) {
            this.aTime = aTime;
            this.aDegree = aDegree;
        }
    }
}

File: src/pooh3/mobi/app/a01_clock/model/TestClock.kt
package pooh3.mobi.app.a01_clock.model

import pooh3.mobi.app.a01_clock.application.GetClockHandsDegree
import sun.jvm.hotspot.utilities.Assert

import java.util.ArrayList

class TestClock {

    private val longHandFixtures = ArrayList<Fixture>()
    private val shortHandFixtures = ArrayList<Fixture>()
    private val shortHandMinFixtures = ArrayList<Fixture>()

    init {
        longHandFixtures.add(Fixture(0, 0))
        longHandFixtures.add(Fixture(15, 90))
        longHandFixtures.add(Fixture(30, 180))
        longHandFixtures.add(Fixture(45, 270))
        //        longHandFixtures.add(new Fixture(60, 360)); // exception

        shortHandFixtures.add(Fixture(0, 0))
        shortHandFixtures.add(Fixture(3, 90))
        shortHandFixtures.add(Fixture(6, 180))
        shortHandFixtures.add(Fixture(9, 270))
        shortHandFixtures.add(Fixture(12, 0))
        shortHandFixtures.add(Fixture(15, 90))
        shortHandFixtures.add(Fixture(18, 180))
        shortHandFixtures.add(Fixture(21, 270))

        shortHandMinFixtures.add(Fixture(0, 0))
        shortHandMinFixtures.add(Fixture(3, 90))
        shortHandMinFixtures.add(Fixture(6, 180))
        shortHandMinFixtures.add(Fixture(9, 90))
        shortHandMinFixtures.add(Fixture(12, 0))
        shortHandMinFixtures.add(Fixture(15, 90))
        shortHandMinFixtures.add(Fixture(18, 180))
        shortHandMinFixtures.add(Fixture(21, 90))
    }

    @Throws(Exception::class)
    fun test() {
        setUp()
        aLongHandTest_minute_to_degree(longHandFixtures)
        aShortHandTest_hour_to_degree(shortHandFixtures)
        aShortHandTest_hour_to_min_degree(shortHandMinFixtures)
        hour10minute10_answer115degree()
    }

    fun setUp() {}

}

private val aShortHandTest_hour_to_degree:
        (shortHandFixtures: ArrayList<Fixture>) -> Unit =
        { list ->
            list.forEach {
                sf ->
                    sf.shortHand
                            .degree(0.longHand).int
                            .also {
                                // side effect
                                println("(hand-degree \n" +
                                        "    (short-hand ${sf.aTimeInt})) \n" +
                                        "> $it") }
                            .then {
                                Assert.that(it == sf.aDegreeInt,
                                        "${sf.aTimeInt} " +
                                                "actual:$it  " +
                                                "expected:${sf.aDegreeInt}")
                            }
            }
        }

private val aShortHandTest_hour_to_min_degree:
        (shortHandMinFixtures: ArrayList<Fixture>) -> Unit =
        { list ->
            list.forEach {
                sf ->
                    sf.shortHand
                            .degree(0.longHand)
                            .minusMin(0.degree)
                            .also {
                                // side effect
                                println("(degree-min (hand-degree \n" +
                                        "                (short-hand ${sf.aTimeInt})) \n" +
                                        "> ${it.int}") }
                            .then {
                                Assert.that(it.int == sf.aDegreeInt,
                                        "${sf.aTimeInt}" +
                                                "actual:${it.int} " +
                                                "expected:${sf.aDegreeInt} ")
                            }
            }
        }

private val aLongHandTest_minute_to_degree:
        (longHandFixtures : ArrayList<Fixture>) -> Unit =
        { list ->
            list.forEach {
                f ->
                    f.longHand.degree.int
                            .also {
                                // side effect
                                println(
                                        "(hand-degree \n" +
                                        "    (long-hand ${f.aTimeInt})) \n" +
                                        "> $it") }
                            .then { degreeInt ->
                                Assert.that(degreeInt == f.aDegreeInt,
                                        "${f.aTimeInt} " +
                                                "actual:$degreeInt " +
                                                "expected:${f.aDegreeInt}")
                            }
            }
        }


private val hour10minute10_answer115degree:
        () -> Unit =
        {
            ("10:10" to GetClockHandsDegree(DegreeCalcService()))
                    .map { (clockStr, getDegree) -> clockStr to getDegree.execute(clockStr) }
                    .also { (clockStr, degree) ->
                        // side effect
                        println("(degree-min (clock-degree \n" +
                                "                (clock-str \"$clockStr\")) \n" +
                                "\n> ${degree.int}") }
                    .then { (clockStr, degree) ->
                        Assert.that(degree.int == 115,
                                "$clockStr actual:${degree.int} expected:" + 115)
                    }
        }



private class Fixture(val aTimeInt: Int, val aDegreeInt: Int)
private val Fixture.longHand: LongHand get() = this.aTimeInt.longHand
private val Fixture.shortHand: ShortHand get() = this.aTimeInt.shortHand
private inline fun <T, R> T.map(block: (T) -> R): R = block(this)
private inline fun <T> T.then(block: (T) -> Unit) { block(this)}

